plane/src/typed_unix_socket/client.rs

```diff
@@ -0,0 +1,209 @@
+        self.event_tx.subscribe()
+    }
+}
+
+async fn handle_connection<MessageToServer, MessageToClient>(
+    stream: UnixStream,
+    mut rx: broadcast::Receiver<WrappedMessage<MessageToServer>>,
+    response_map: Arc<DashMap<String, oneshot::Sender<MessageToClient>>>,
+    event_tx: broadcast::Sender<MessageToClient>,
+) -> Result<(), Box<dyn std::error::Error + Send + Sync>>
+where
+    MessageToServer: Send + Sync + 'static + Clone + Debug + Serialize + for<'de> Deserialize<'de>,
+    MessageToClient: Send + Sync + 'static + Clone + Debug + Serialize + for<'de> Deserialize<'de>,
+{
+    let (read_half, write_half) = tokio::io::split(stream);
+    let reader = BufReader::new(read_half);
+    let writer = BufWriter::new(write_half);
+
+    let mut lines = reader.lines();
+    let mut writer = writer;
+
+    // Task to handle receiving messages
+    let recv_task = tokio::spawn({
+        let event_tx = event_tx.clone();
+        let response_map = Arc::clone(&response_map);
+        async move {
+            loop {
+                match lines.next_line().await {
+                    Ok(Some(line)) => {
+                        let msg: WrappedMessage<MessageToClient> = match serde_json::from_str(&line)
+                        {
+                            Ok(msg) => msg,
+                            Err(e) => {
+                                tracing::error!("Error deserializing message: {}", e);
+                                continue;
+                            }
+                        };
+                        match msg {
+                            WrappedMessage {
+                                id: Some(id),
+                                message,
+                            } => {
+                                if let Some((_, tx)) = response_map.remove(&id) {
+                                    if let Err(e) = tx.send(message) {
+                                        tracing::error!("Error sending response: {:?}", e);
+                                    }
+                                } else {
+                                    tracing::error!("No sender found for response ID: {:?}", id);
+                                }
+                            }
+                            WrappedMessage { id: None, message } => {
+                                if let Err(e) = event_tx.send(message) {
+                                    tracing::error!("Error sending event: {}", e);
+                                }
+                            }
+                        }
+                    }
+                    Ok(None) => {
+                        tracing::error!("Connection closed by server");
+                    }
+                    Err(e) => {
+                        tracing::error!("Error reading line: {}", e);
+                    }
+                }
+            }
+        }
+    });
+
+    // Task to handle sending messages
+    let send_task = tokio::spawn(async move {
+        loop {
```

Comment:
I don't think this loop ever terminates, even after a reconnect (left a similar comment on the server with a suggestion)